File: tape-read.pl
#!/usr/bin/perl
#
# tape-read.pl
#
# read data from a Compact Cassette
#
# (c) windytan (Oona Räisänen)
#
# ISC license
#

use warnings;
$|++;

my %conf;
open(IN,"ctape.conf") or die($!);
for (<IN>) {
  chomp;
  $conf{$1} = $2 if (/^(\S+) (.+)/);
}
close(IN);

open(IN,"sox -q ".$conf{'device'}." -t .raw -c 1 -r 44100 -b 16 -e signed-integer - |");

# calibrate polarity
# (wait for 50 repetitions of (nppp) or (pnnn))
my $polty = 0;
while ( (not eof(IN)) && $polty == 0 ) {
  read(IN,$samp,2);

  $samp = unpack("s",$samp);
  $slen++;
  if (($prevsamp // 0) * $samp < 0) {
    $calstring .= ($prevsamp > 0 ? "p" : "n") x round($slen / $conf{'bitlen'});
    $calstring = substr($calstring,-150) if (length($calstring) > 150);

    if (round($slen / $conf{'bitlen'}) > 0) {
      if ($calstring =~ /(nppp){30}/) {
        $polty = 1;
        last;
      }
      if ($calstring =~ /(pnnn){30}/) {
        $polty = -1;
        last;
      }
    }
    $slen = 0;
  }
  $prevsamp = $samp;
}

# read data

open(G,"|sox -q -t .raw -c 1 -r 44100 -b 16 -e signed-integer - g.wav");#sinc -10640|");
my $prev_c  = 0;
my $bitreg  = 0;
my $bytereg = 0;
while (not eof(IN)) {
  read(IN,$a,2);

  $c = $polty * unpack("s",$a);
  print G pack("s",$c);
  $len ++;
  if ($prev_c > 0 && $c <= 0) {
    bit($len > $conf{'bitlen'}*1.5);
    $len = 0;
  }
  $prev_c = $c;
}
close(IN);
close(G);



sub bit {

  $bitreg = (($bitreg << 1) & 0x3FF) + $_[0];

  if (not $bitsync) {
    if (($bitreg >> 9) && not ($bitreg & 1)) {
      $bitsync  = 1;
      $bitcount = 0;
    } else {
      $bitsync = $bytereg = 0;
    }

  } else {

    if (++$bitcount == 10) {
      $bitcount=0;
      if (not (($bitreg >> 9) && not ($bitreg & 1))) {
        die if ($bytesync);
        $bitsync = $bytesync = $bytereg = 0;
      } elsif ($bytesync) {
        print chr(($bitreg >> 1) & 0xFF);
      } else {
        $bytereg = (($bytereg << 8) + (($bitreg >> 1) & 0xFF)) & 0xFFFFFFFF;
        $bytesync = 1 if ($bytereg == 0x08070504);
      }
    }
  }

}

sub round {
  int($_[0] + .5);
}

File: tape-read.rb
#!/usr/bin/ruby -w
# ctape: read data from a Compact Cassette
# (c) windytan / Oona Räisänen
# ISC license

require 'yaml'

conf = YAML::load(File.open('config.yml'))
$bitlen = conf['bitlen']
$device = conf['device']

$bitreg  = 0
$bytereg = 0
$bitsync = false
$bytesync = false

def bit(value)
  $bitreg = (($bitreg << 1) & 0x3FF) + (value ? 1 : 0)

  if !$bitsync
    if ($bitreg >> 9 > 0) && ($bitreg & 1) == 0
      $bitsync  = true
      $bitcount = 0
    else
      $bitsync = false
      $bytereg = 0
    end
  else
    $bitcount += 1
    if $bitcount == 10
      $bitcount = 0
      if !(($bitreg >> 9 > 0) && ($bitreg & 1) == 0)
        abort if $bytesync
        $bitsync  = false
        $bytesync = false
        $bytereg  = 0
      elsif $bytesync
        print((($bitreg >> 1) & 0xFF).chr)
        $stdout.flush
      else
        $bytereg = (($bytereg << 8) + (($bitreg >> 1) & 0xFF)) & 0xFFFFFFFF
        $bytesync = true if $bytereg == 0x08070504
      end
    end
  end
end


$sox = IO.popen('sox -q '+$device+' -t .raw -r 44100 -c 1 -b 16 -e signed-integer -','r')

# Calibrate polarity
# (wait for 50 repetitions of nppp or pnnn)

polarity  = 0
prevsamp  = 0
slen      = 0
calstring = ""
until $sox.eof? or polarity != 0
  sample = $sox.read(2).unpack('s')[0]
  slen += 1
  if prevsamp * sample < 0
    calstring += (prevsamp > 0 ? "p" : "n") * (slen.to_f / $bitlen).round
    calstring = calstring.slice(-150,150) if calstring.length > 150

    if (slen.to_f / $bitlen).round > 0
      if calstring =~ /(nppp){30}/
        polarity = 1
        break
      end
      if calstring =~ /(pnnn){30}/
        polarity = -1
        break
      end
    end
    slen = 0
  end
  prevsamp = sample
end

# Read data

prevsamp  = 0
curlen  = 0
until $sox.eof?
  sample = $sox.read(2).unpack('s')[0] * polarity
  curlen += 1
  if prevsamp > 0 && sample <= 0
    bit(curlen > $bitlen*1.5)
    curlen = 0
  end
  prevsamp = sample
end

$sox.close
